[EXTERNAL PR] TTNN Roll Op Support

Bhuvanesh194Sankar opened the pull request:
### #12157 
Link to Github Issue : https://github.com/tenstorrent/tt-metal/issues/12157

### Problem description
Need support for ttnn roll() op.

### What's changed
Implemented roll op (ttnn.roll) as a composite op using the ttnn.slice and ttnn.concat, and added pytest .

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI [passes](https://github.com/tenstorrent/tt-metal/actions/runs/15045391672/job/42312144550)
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

Bhuvanesh194Sankar committed:
Added Roll operation

Bhuvanesh194Sankar committed:
Add Roll Operation

Bhuvanesh194Sankar commented:
Hello team, can someone kindly review this PR ?

bbradelTT commented on tests/ttnn/unit_tests/test_roll.py:
```diff
@@ -0,0 +1,53 @@
+@pytest.mark.parametrize(
+    "input_tensor, shifts, dim",
+    [
+        ([[1, 2], [3, 4], [5, 6], [7, 8]], 1, 0),
```
Is dim being None supported?
- If it is supported, please add a test.
- If it isn't supported, should it be supported?
  - If it should be, then the code needs to be added (in this or a subsequent PR).
  - If it shouldn't be supported, please note that in the doc string.

Bhuvanesh194Sankar replied:
Will add support for dim being None in this pr itself.

Bhuvanesh194Sankar replied:
@bbradelTT I have added the None dim support can you kindly review it ?

bbradelTT replied:
Where are the tests for the None dim?

bbradelTT commented on ttnn/cpp/ttnn/operations/data_movement/roll/roll.cpp:
```diff
@@ -0,0 +1,69 @@
+
+ttnn::Tensor RollOperation::invoke(
+    const ttnn::Tensor& input_tensor, const std::vector<int>& shifts, const std::vector<int>& dims) {
+    TT_FATAL(!shifts.empty() && shifts.size() == dims.size(), "Roll expects shifts and dims to have the same length");
```
Please print out values in the error statement.

bbradelTT commented on ttnn/cpp/ttnn/operations/data_movement/roll/roll.cpp:
```diff
@@ -0,0 +1,69 @@
+    int num_dims = size.rank();
+
+    for (int dim : dims) {
+        TT_FATAL(dim >= -num_dims && dim < num_dims, "Invalid dimension index.");
```
Please print out values in the error statement.

Bhuvanesh194Sankar commented:
@bbradelTT Thank you for your comments. I will work on them and update the changes at the earliest

Bhuvanesh194Sankar commented on ttnn/cpp/ttnn/operations/data_movement/roll/roll.cpp:
```diff
@@ -0,0 +1,69 @@
+        int dim = dims[i];
+
+        int shift_size = input_tensor.get_logical_shape()[dim];
+        adjusted_shifts[i] = ((shift % shift_size) + shift_size) % shift_size;
```
@bbradelTT 
adjusted_shifts[i] = (shift + shift_size) % shift_size; -> This code snippet that you have suggested will work on positive shift, 
adjusted_shifts[i] = ((shift % shift_size) + shift_size) % shift_size; -> This will handle both postive and negative shifts aswell.
So to handle the negative shifts our code snippet is necessary.

Commit:
Added values to print in error message

ntarafdar commented on tests/ttnn/unit_tests/test_roll.py:
```diff
@@ -0,0 +1,53 @@
+def test_roll(device, input_tensor, shifts, dim):
+    tensor = torch.tensor(input_tensor, dtype=torch.float32)
+    ttnn_tensor = ttnn.from_torch(tensor)
+    ttnn_tensor = ttnn.to_layout(ttnn_tensor, ttnn.ROW_MAJOR_LAYOUT)
```
can we test for TILE and other dtypes (bfloat16 in particular) as well. 

Also for TILE can we try shapes that are not divisible by 32 for the last two dims to see how this would handle padding. 

It should be okay as it uses slice and concat underneath the hood which has this support. 

Can we also try L1 memory and diff sharding configs as well. 

And lastly can we test the program cache, (use_program_cache and have multiple runs)

ntarafdar reviewed (changes requested):
just a small comment on testing.

Commit:
Added Tests and none dim support

Bhuvanesh194Sankar commented:
@ntarafdar I have done the suggested changes, can you kindly review it again ?

Bhuvanesh194Sankar commented on tests/ttnn/unit_tests/test_roll.py:
```diff
@@ -33,21 +33,193 @@
     ],
 )
-def test_roll(device, input_tensor, shifts, dim):
+def test_roll_without_dim(device, input_tensor, shifts):
```
@bbradelTT here is the test for roll without dim

bbradelTT replied:
Thanks.

Commit:
Removed Print statements in the tests

Bhuvanesh194Sankar commented:
@bbradelTT removed the print statements from the test fucntions can you kindly review it.

bbradelTT commented on ttnn/cpp/ttnn/operations/data_movement/roll/roll.cpp:
```diff
@@ -0,0 +1,219 @@
+    return result;
+}
+
+ttnn::Tensor RollOperation::invoke(const ttnn::Tensor& input_tensor, const int& shift) {
```
This is too much code duplication. There should be one method that does the operations (like in the first commit).
- Either this is an internal method and all three invoke methods call that one.
- Or it is the initial invoke method with the other two invoke methods (this one and the one below) calling the initial invoke method.

bbradelTT commented on ttnn/cpp/ttnn/operations/data_movement/roll/roll_pybind.cpp:
```diff
@@ -0,0 +1,67 @@
+        Args:
+            input_tensor: A tensor whose elements will be rolled.
+            shifts: The number of places by which elements are shifted. Can be an integer or a list of integers (one per dimension).
+            dim: The dimension(s) along which to roll. Must have the same length as `shifts` if a list is provided.
```
This is not clear if the list is for shifts or dim. 
The end should probably be changed to
"If `shifts` is a list, then `dim` must be a list of the same length as `shifts`."

Commit:
Updated the pybind docs and removed the reference throughout the specified datatypes and worked on duplication of the code

Bhuvanesh194Sankar commented:
Kindly share if there are any other review comments. If not, could anyone please trigger the CIs for the PR? Thank you. Also, @ntarafdar  , your approval might be blocking the PR. Could you approve the PR, if no further comments from your side?

bbradelTT commented:
> @bbradelTT Kindly share if there are any other review comments. If not, could you please trigger the CIs for the PR? Thank you. Also, @ntarafdar , your approval might be blocking the PR. Could you approve the PR, if no further comments from your side?

Thanks @Bhuvanesh194Sankar the changes look good.

I think you still need to address the comment from @ntarafdar about the tests https://github.com/tenstorrent/tt-metal/pull/19676/files#r2039494468 and then he can trigger the CIs.

Bhuvanesh194Sankar commented:
> > @bbradelTT Kindly share if there are any other review comments. If not, could you please trigger the CIs for the PR? Thank you. Also, @ntarafdar , your approval might be blocking the PR. Could you approve the PR, if no further comments from your side?
> 
> Thanks @Bhuvanesh194Sankar the changes look good.
> 
> I think you still need to address the comment from @ntarafdar about the tests https://github.com/tenstorrent/tt-metal/pull/19676/files#r2039494468 and then he can trigger the CIs.

Thank you @bbradelTT.  Regarding the comment from @ntarafdar, I have already addressed it and added the suggested tests in the test_roll.py test file.

Bhuvanesh194Sankar commented:
@bbradelTT @ntarafdar The tests for the different sharding configs were not included because this roll op makes use of slice and concat, the slice supports sharding but concat has some constraints on the sharding, which throws the fatals, the fatals in the code is listed below as the images for your reference, these fatals are in the file 
tt-metal/ttnn/cpp/ttnn/operations/data_movement/concat/device/concat_device_operation.cpp

![image](https://github.com/user-attachments/assets/7b796515-898f-48d9-a165-4f9edbc875a7)

![image](https://github.com/user-attachments/assets/de9475ba-b7c9-429c-8b74-a5b5a6119f8f)

So other than that I have addressed all the other comments for test suggested by @ntarafdar which was in the test_roll.py test python file.

Bhuvanesh194Sankar commented:
@bbradelTT @ntarafdar  Could some one review this ?

Bhuvanesh194Sankar commented:
@ntarafdar  Thank you for the approval, the CI runs triggered yesterday seemed fine. I have rebased the branch to the latest main. Could you please trigger the CI runs again? If they pass can we proceed with merge?

Commit:
Resolved merge conflict

Commit:
Resolved the merge conflicts

Bhuvanesh194Sankar commented:
@amorrisonTT Could you kindly review this PR  ?

Commit:
Updated the pybind

Bhuvanesh194Sankar commented:
@amorrisonTT Could you kindly review this PR ?

amorrisonTT commented on ttnn/cpp/ttnn/operations/data_movement/roll/roll.cpp:
```diff
@@ -0,0 +1,110 @@
+namespace ttnn::operations::data_movement {
+
+ttnn::Tensor RollOperation::invoke(
+    const ttnn::Tensor& input_tensor, const ttnn::SmallVector<int>& shift, const ttnn::SmallVector<int>& dim) {
```
I think you need to rename the input argument `dim` since it gets shadowed in the function body. I would suggest:  `input_dims`

Commit:
Addressed the review comments

Bhuvanesh194Sankar commented:
@amorrisonTT I have done the suggested changes, could you kindly review it again ?

Bhuvanesh194Sankar commented:
@amorrisonTT We can proceed with the CI runs .

tt-rkim commented:
running https://github.com/tenstorrent/tt-metal/actions/runs/14910320925/job/41884018121?pr=19676

Bhuvanesh194Sankar commented:
@tt-rkim it seems five checks were failing. What should we do next ?

Bhuvanesh194Sankar commented:
@tt-rkim I have rebased the branch to the latest main, can we go for another CI run?

Bhuvanesh194Sankar commented:
@tt-rkim I have rebased the branch to the latest main. Could you please trigger the CI runs again ?

tt-rkim commented:
APC here https://github.com/tenstorrent/tt-metal/actions/runs/14995088226/job/42132248644?pr=19676

sminakov-tt commented on ttnn/cpp/ttnn/operations/data_movement/roll/roll.cpp:
```diff
@@ -0,0 +1,113 @@
+ttnn::Tensor RollOperation::invoke(
+    const ttnn::Tensor& input_tensor, const ttnn::SmallVector<int>& shift, const ttnn::SmallVector<int>& input_dims) {
+    std::vector<int> shifts(shift.begin(), shift.end());
+    std::vector<int> dims(input_dims.begin(), input_dims.end());
```
It seems like this conversion from SmallVector to std::vector is redundant, can we use SmallVector throughout this function?

Bhuvanesh194Sankar committed:
Resolved the redundancy

tt-rkim commented:
Note that other people can run it as well, and someone already triggered it. Please refer to the statuses below on this page, before rebasing again, as that will erase it.
